### PR TITLE
[Bugfix] Update run_batch.py to handle larger numbers of batches

### LINKED ISSUE #6154: [Feature]: expose the tqdm progress bar to enable logging the progress
### 🚀 The feature, motivation and pitch

I am running vLLM on aws batch on large datasets and would like to see the progress of the jobs, this is currently not possible. 
however [tqdm-loggable](https://pypi.org/project/tqdm-loggable/) does enable this, but then I would need to swap the tqdm driver in the llm_engine, this is however not exposed.
so my request would be either to move to tqdm-loggable since it automatically switches given a interactive vs non-interactive environment. Or expose tqdm such that the user can swap it themselves.

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by w013nad**
vllm engine appears to error out somewhere between 20k-100k async requests. Refactor to only allow a maximum of 10k concurrent requests with Semaphore. Also added in a tqdm progress bar. Note that this was created under 0.5.0.post1, I am unable to test integration with some of the newer changes, but it should work.

It would be nice to suppress the request output as well and just have the option to have tqdm only but I was unable to figure out how to make that work.

Note that this is my first PR for this project so I'm not entirely sure how all of this works.

FIX #6154

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by wuisawesome** [comment]
@w013nad do you mind sharing the error you got without this PR? 

I think we should definitely support this many requests, but wonder if there's an underlying issue that we should fix.

File: vllm/entrypoints/openai/run_batch.py [hunk 1]
```diff
@@ -100,6 +101,28 @@ async def run_request(chat_serving: OpenAIServingChat,
...
+        async with sem:
+            
+            return await run_request(openai_serving_chat, request)
+    
+    response_futures: List[Awaitable[BatchRequestOutput]] = []
+    
+    # Read file contents asynchronously
+    file_contents = (await read_file(input_file)).strip().split("\n")
+    total_lines = len(file_contents)
+
```
Comment (wuisawesome) [inline vllm/entrypoints/openai/run_batch.py:119]: Let's figure out whether the engine queuing should be fixed instead, but if we decide to limit the requests we submit, I wonder if we should just not create the asyncio requests instead.

Something like

```
finished_requests = []
response_futures: List[Awaitable[BatchRequestOutput]] = []

for request_json in tqdm(file_contents, total=total_lines, desc="Running requests"):
        request = BatchRequestInput.model_validate_json(request_json.strip())
        response_futures.append(run_request_with_semaphore(request))
        if len(response_futures) >= MAX_CONCURRENT_REQUESTS:
            recently_finished, response_futures = await asyncio.wait(response_futures)
            finished_requests.extend(recently_finished)

finished_requests.extend(await asyncio.gather(*response_futures))
```

This way we can more accurately track the progress of the job.

**Comment by cadedaniel** [comment]
FYI there are ongoing refactors to move engine/api server/worker into own processes. This should enable much higher QPS.

**Comment by w013nad** [comment]
I'm doing a RAG test running 4500 questions through with k=[0,1,2,3,5,6,7,8,9,10,11,12,13,14,15,17,20,25, 50, 75] chunks. The total file size is 1.2GB and 95k requests. I did another test which also had 95k requests, but the queries were much shorter, and it worked fine. This indicates it's not the number of requests but the amount of text I'm sending in. The other batch file size was around 200 MB.


```
INFO 06-23 22:52:10 async_llm_engine.py:564] Received request cmpl-f96570de404141e09427685b17c752c4: prompt: '<|im_start', params: SamplingParams(n=1, best_of=1, presence_penalty=0.0, frequency_penalty=0.0, repetition_penalty=1.0, temperature=0.01, top_p=1.0, top_k=-1, min_p=0.0, seed=None, use_beam_search=False, length_penalty=1.0, early_stopping=False, stop=[], stop_token_ids=[], include_stop_str_in_output=False, ignore_eos=False, max_tokens=1000, min_tokens=0, logprobs=None, prompt_logprobs=None, skip_special_tokens=True, spaces_between_special_tokens=True, truncate_prompt_tokens=None), prompt_token_ids: [151644, 8948, 198, 2610, 525, 264, 10950, 17847, 13, 151645], lora_request: None.
ERROR 06-23 22:54:02 async_llm_engine.py:535] Engine iteration timed out. This should never happen!
ERROR 06-23 22:54:02 async_llm_engine.py:52] Engine background task failed
ERROR 06-23 22:54:02 async_llm_engine.py:52] Traceback (most recent call last):
ERROR 06-23 22:54:02 async_llm_engine.py:52]   File "/usr/local/lib/python3.10/dist-packages/vllm/engine/async_llm_engine.py", line 506, in engine_step
ERROR 06-23 22:54:02 async_llm_engine.py:52]     request_outputs = await self.engine.step_async()
ERROR 06-23 22:54:02 async_llm_engine.py:52]   File "/usr/local/lib/python3.10/dist-packages/vllm/engine/async_llm_engine.py", line 235, in step_async
ERROR 06-23 22:54:02 async_llm_engine.py:52]     output = await self.model_executor.execute_model_async(
ERROR 06-23 22:54:02 async_llm_engine.py:52]   File "/usr/local/lib/python3.10/dist-packages/vllm/executor/distributed_gpu_executor.py", line 166, in execute_model_async
ERROR 06-23 22:54:02 async_llm_engine.py:52]     return await self._driver_execute_model_async(execute_model_req)
ERROR 06-23 22:54:02 async_llm_engine.py:52]   File "/usr/local/lib/python3.10/dist-packages/vllm/executor/multiproc_gpu_executor.py", line 149, in _driver_execute_model_async
ERROR 06-23 22:54:02 async_llm_engine.py:52]     return await self.driver_exec_model(execute_model_req)
ERROR 06-23 22:54:02 async_llm_engine.py:52] asyncio.exceptions.CancelledError
ERROR 06-23 22:54:02 async_llm_engine.py:52]
ERROR 06-23 22:54:02 async_llm_engine.py:52] During handling of the above exception, another exception occurred:
ERROR 06-23 22:54:02 async_llm_engine.py:52]
ERROR 06-23 22:54:02 async_llm_engine.py:52] Traceback (most recent call last):
ERROR 06-23 22:54:02 async_llm_engine.py:52]   File "/usr/lib/python3.10/asyncio/tasks.py", line 456, in wait_for
ERROR 06-23 22:54:02 async_llm_engine.py:52]     return fut.result()
ERROR 06-23 22:54:02 async_llm_engine.py:52] asyncio.exceptions.CancelledError
ERROR 06-23 22:54:02 async_llm_engine.py:52]
ERROR 06-23 22:54:02 async_llm_engine.py:52] The above exception was the direct cause of the following exception:
ERROR 06-23 22:54:02 async_llm_engine.py:52]
ERROR 06-23 22:54:02 async_llm_engine.py:52] Traceback (most recent call last):
ERROR 06-23 22:54:02 async_llm_engine.py:52]   File "/usr/local/lib/python3.10/dist-packages/vllm/engine/async_llm_engine.py", line 42, in _log_task_completion
ERROR 06-23 22:54:02 async_llm_engine.py:52]     return_value = task.result()
ERROR 06-23 22:54:02 async_llm_engine.py:52]   File "/usr/local/lib/python3.10/dist-packages/vllm/engine/async_llm_engine.py", line 532, in run_engine_loop
ERROR 06-23 22:54:02 async_llm_engine.py:52]     has_requests_in_progress = await asyncio.wait_for(
ERROR 06-23 22:54:02 async_llm_engine.py:52]   File "/usr/lib/python3.10/asyncio/tasks.py", line 458, in wait_for
ERROR 06-23 22:54:02 async_llm_engine.py:52]     raise exceptions.TimeoutError() from exc
ERROR 06-23 22:54:02 async_llm_engine.py:52] asyncio.exceptions.TimeoutError
INFO 06-23 22:54:03 async_llm_engine.py:167] Aborted request cmpl-e0d8265bcc46475d944cd1b0be30b316.
```

After this, I get 95k lines of 
```
INFO 06-23 22:54:06 async_llm_engine.py:167] Aborted request cmpl-f96570de404141e09427685b17c752c4.
```
and then it just shuts itself down.

**Comment by wuisawesome** [comment]
gotcha, this makes sense. @w013nad do you mind making the change to the streaming approach and let's just get this merged in the front end?

**Pull request closed without merging**